File: packages/python/src/agent_wallet/core/clients/privy.py
"""Privy HTTP client."""

from __future__ import annotations

import base64
import json
import time
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from agent_wallet.core.errors import PrivyAuthError, PrivyRateLimitError, PrivyRequestError


class PrivyClient:
    def __init__(
        self,
        *,
        app_id: str,
        app_secret: str,
        retries: int = 2,
        sleep: callable | None = None,
    ) -> None:
        self._app_id = app_id
        self._app_secret = app_secret
        self._base_url = "https://api.privy.io"
        self._retries = retries
        self._sleep = sleep or (lambda seconds: time.sleep(seconds))

    def get_wallet(self, wallet_id: str) -> dict[str, Any]:
        return self._request("GET", f"/v1/wallets/{wallet_id}")

    def rpc(
        self,
        wallet_id: str,
        method: str,
        params: dict[str, Any],
        authorization_signature: str | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            f"/v1/wallets/{wallet_id}/rpc",
            {"method": method, "params": params},
            authorization_signature=authorization_signature,
        )

    def raw_sign(
        self, wallet_id: str, params: dict[str, Any], authorization_signature: str | None = None
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            f"/v1/wallets/{wallet_id}/raw_sign",
            {"params": params},
            authorization_signature=authorization_signature,
        )

    def _request(
        self,
        method: str,
        path: str,
        body: dict[str, Any] | None = None,
        *,
        authorization_signature: str | None = None,
    ) -> dict[str, Any]:
        attempt = 0
        while True:
            req = Request(
                f"{self._base_url}{path}",
                method=method,
                headers=self._headers(authorization_signature),
                data=json.dumps(body).encode("utf-8") if body else None,
            )
            try:
                with urlopen(req) as response:
                    status = response.status
                    payload = _read_json(response)
            except HTTPError as exc:
                status = exc.code
                try:
                    payload = json.loads(exc.read().decode("utf-8"))
                except Exception:
                    payload = {}
            except URLError as exc:
                raise PrivyRequestError(f"Privy request network error: {exc}") from exc

            if status == 429:
                if attempt >= self._retries:
                    raise PrivyRateLimitError("Privy rate limit exceeded")
                attempt += 1
                self._sleep(_backoff_seconds(attempt))
                continue

            if status >= 400:
                message = _extract_error(payload) or f"Privy request failed with status {status}"
                if status in (401, 403):
                    raise PrivyAuthError(message)
                raise PrivyRequestError(message)

            return payload

    def _headers(self, authorization_signature: str | None) -> dict[str, str]:
        raw = f"{self._app_id}:{self._app_secret}".encode()
        auth = base64.b64encode(raw).decode("utf-8")
        headers = {
            "accept": "application/json",
            "content-type": "application/json",
            "authorization": f"Basic {auth}",
            "privy-app-id": self._app_id,
            "user-agent": "agent-wallet-python",
        }
        if authorization_signature:
            headers["privy-authorization-signature"] = authorization_signature
        return headers


def _read_json(response) -> dict[str, Any]:
    try:
        return json.loads(response.read().decode("utf-8"))
    except Exception:
        return {}


def _extract_error(payload: dict[str, Any]) -> str | None:
    if "error" in payload and isinstance(payload["error"], dict):
        message = payload["error"].get("message")
        if isinstance(message, str):
            return message
    if isinstance(payload.get("message"), str):
        return payload["message"]
    return None


def _backoff_seconds(attempt: int) -> float:
    return min(1.0, 0.2 * attempt)

File: packages/typescript/src/core/clients/privy.ts
import { PrivyAuthError, PrivyRateLimitError, PrivyRequestError } from '../errors.js'

export type PrivyClientConfig = {
  appId: string
  appSecret: string
  retries?: number
  sleep?: (ms: number) => Promise<void>
}

export type PrivyRpcMethod =
  | 'personal_sign'
  | 'eth_signTransaction'
  | 'eth_signTypedData_v4'
  | 'raw_sign'

export type PrivyRpcParams = Record<string, unknown>

export type PrivyRpcResponse = {
  data: {
    signature?: string
    signed_transaction?: string
  }
}

export type PrivyRawSignParams = {
  hash?: string
  bytes?: string
  encoding?: string
  hash_function?: string
}

export class PrivyClient {
  private readonly appId: string
  private readonly appSecret: string
  private readonly baseUrl: string
  private readonly retries: number
  private readonly sleep: (ms: number) => Promise<void>

  constructor(config: PrivyClientConfig) {
    this.appId = config.appId
    this.appSecret = config.appSecret
    this.baseUrl = 'https://api.privy.io'
    this.retries = config.retries ?? 2
    this.sleep = config.sleep ?? ((ms) => new Promise((resolve) => setTimeout(resolve, ms)))
  }

  async getWallet(walletId: string): Promise<{ address: string; chainType: string }> {
    const result = await this.request('GET', `/v1/wallets/${walletId}`)
    const data = readDataObject(result)
    if (!data?.address) {
      throw new PrivyRequestError('Privy wallet response missing address')
    }
    return { address: data.address, chainType: data.chain_type ?? '' }
  }

  async rpc(
    walletId: string,
    method: PrivyRpcMethod,
    params: PrivyRpcParams,
    options?: PrivyRequestOptions,
  ): Promise<PrivyRpcResponse> {
    return this.request(
      'POST',
      `/v1/wallets/${walletId}/rpc`,
      {
      method,
      params,
      },
      options,
    ) as Promise<PrivyRpcResponse>
  }

  async rawSign(
    walletId: string,
    params: PrivyRawSignParams,
    options?: PrivyRequestOptions,
  ): Promise<PrivyRpcResponse> {
    return this.request(
      'POST',
      `/v1/wallets/${walletId}/raw_sign`,
      {
        params,
      },
      options,
    ) as Promise<PrivyRpcResponse>
  }

  private async request(
    method: 'GET' | 'POST',
    path: string,
    body?: unknown,
    options?: PrivyRequestOptions,
  ): Promise<unknown> {
    let attempt = 0
    while (true) {
      const response = await fetch(`${this.baseUrl}${path}`, {
        method,
        headers: this.buildHeaders(options),
        body: body ? JSON.stringify(body) : undefined,
      })

      if (response.status === 429) {
        if (attempt >= this.retries) {
          throw new PrivyRateLimitError('Privy rate limit exceeded')
        }
        attempt += 1
        await this.sleep(backoffMs(attempt))
        continue
      }

      const payload = await readJson(response)
      if (!response.ok) {
        const message =
          extractErrorMessage(payload) ?? `Privy request failed with status ${response.status}`
        if (response.status === 401 || response.status === 403) {
          throw new PrivyAuthError(message)
        }
        throw new PrivyRequestError(message)
      }

      return payload
    }
  }

  private buildHeaders(options?: PrivyRequestOptions): Record<string, string> {
    const auth = Buffer.from(`${this.appId}:${this.appSecret}`).toString('base64')
    const headers: Record<string, string> = {
      'content-type': 'application/json',
      authorization: `Basic ${auth}`,
      'privy-app-id': this.appId,
    }
    if (options?.authorizationSignature) {
      headers['privy-authorization-signature'] = options.authorizationSignature
    }
    return headers
  }
}

export type PrivyRequestOptions = {
  authorizationSignature?: string
}

function backoffMs(attempt: number): number {
  return Math.min(1000, 200 * attempt)
}

async function readJson(response: Response): Promise<unknown> {
  try {
    return await response.json()
  } catch {
    return undefined
  }
}

function readDataObject(payload: unknown): { address?: string; chain_type?: string } | null {
  if (!payload || typeof payload !== 'object') return null
  const record = payload as Record<string, unknown>
  if (record.data && typeof record.data === 'object') {
    return record.data as { address?: string; chain_type?: string }
  }
  if ('address' in record || 'chain_type' in record) {
    return record as { address?: string; chain_type?: string }
  }
  return null
}

function extractErrorMessage(payload: unknown): string | undefined {
  if (!payload || typeof payload !== 'object') return undefined
  const record = payload as Record<string, unknown>
  if (record.error && typeof record.error === 'object') {
    const error = record.error as Record<string, unknown>
    if (typeof error.message === 'string') return error.message
  }
  if (typeof record.message === 'string') return record.message
  return undefined
}
